File: Koosha/Arduino/src/edu/wpi/first/wpilibj/templates/Multiplexer.java
/*
 * To change this template, choose Tools | Templates
 * and open the template in the editor.
 */
package edu.wpi.first.wpilibj.templates;

import edu.wpi.first.wpilibj.*;

/**
 *
 * @author Koosha Jahani
 */
public class Multiplexer extends SensorBase
{
    private final int kDigitalModule = 1;
    private final int kAddress = 0x64; // address is 100
    private byte hiByte;
    private byte loByte;
    private boolean successful;
    
    private I2C m_I2C;
    
    public Multiplexer()
    {
        m_I2C = new I2C(DigitalModule.getInstance(kDigitalModule), kAddress);
    }
     
/*
 * reads the arduino and recieves 2 bytes to represent the digital inputs attached to it
 * saves the 2 bytes to 'loByte' and 'hiByte'
 * return: true if and only if it's successful
 */
    public boolean readArduino()
    {
        byte[] buffer = {0,0,0};
        
        successful = !m_I2C.transaction(null, 0, buffer, 2); // doesn't send anything and recieves 2 bytes of data; successful==true if transaction is succesful
        DriverStationLCD.getInstance().println(DriverStationLCD.Line.kUser1, 1, "success=" + successful + "buffer=" + buffer);
        DriverStationLCD.getInstance().println(DriverStationLCD.Line.kUser2, 1, "buffer=" + buffer);
        DriverStationLCD.getInstance().updateLCD();
        hiByte = buffer[1];
        loByte = buffer[0];
        
        return successful;
    }
    
    
/*
 * reads the value of the digital input of the specified channel
 * param 1: the channel to be read
 * param 2: a bool varible to hold the sucess of the function
 * return: value of the specified digital input
 */
    public boolean getInput(int channelID, boolean[] success)
    {
        byte temp;
        
        if(Math.abs(channelID-(13/2))-(13/2)>0) // returns true if less than zero or greater than thirteen
        {
            DriverStationLCD.getInstance().println(DriverStationLCD.Line.kUser2, 1, "ERROR; invalid ID must be between 0 and 13");
            success[0] = false;
            return false;
        }
        else if(channelID < 8)
        {
            temp = (byte) (loByte & (1 << channelID));
            success[0] = false;
            return temp!=0;
        }
        else
        {
            temp = (byte) (hiByte & (1 << channelID-8));
            success[0] = false;
            return temp!=0;
        }
    }
}

File: Koosha/2012 Audrino Program(c++)/Arduino_program/Multiplexer.cpp
#include "WPILib.h"
#include "Multiplexer.h"

Multiplexer::Multiplexer()
{
	 DigitalModule *module = DigitalModule::GetInstance(1);
	if (module)
	{
		m_i2c = module->GetI2C(kAddress);
	}
}

/**
 * Destructor.
 */
Multiplexer::~Multiplexer()
{
	delete m_i2c;
	m_i2c = NULL;
}

/*
 * reads the arduino and recieves 2 bytes to represent the digital inputs attached to it
 * saves the 2 bytes to 'loByte' and 'hiByte'
 * return: true if unseccessful
 */
bool Multiplexer::readArduino()
{
	UINT8 buffer[3];
	bool result;

	result = m_i2c ->Transaction(0,0,buffer,2);
//	printf(" Transaction result=%d, buffer[]=", result);
	dataRead.loByte = buffer [0];
	dataRead.hiByte = buffer [1];
	dataRead.successful = result;
	if(!result)
	{
		//printf("%x, %x\n", buffer[0], buffer[1]);
	}
	else
	{
		//printf("ERROR: can not recieve data from arduino");
	}

	return result;
}

/*
 * reads the value of the digital input of the specified channel
 * param 1: the channel to be read
 * param 2: a bool varible to hold the sucess of the function
 * return: value of the specified digital input
 */
bool Multiplexer::getInput(int channelID, bool *success)
{
	uint8_t temp;
	
	if(channelID>13 || channelID<0)
	{
		printf("ERROR: invalid ID must be between 0 and 13");
		*success = false;
		return false;
	}
	*success = true;
	
	if (channelID < 8)
	{
		temp = dataRead.loByte &(1 << channelID);
		return !temp == 0;
	}
	else
	{
		temp=dataRead.hiByte &(1 << channelID-8);
		return !temp == 0;
	}
}
